beard/clustering/blocking.py

```diff
@@ -97,58 +150,41 @@ def _validate(self, X, blocks):
         """Fit base clustering estimators on X."""
+        def _grouper(size, seq):
+            return (seq[pos:pos + size] for pos in xrange(0, len(seq), size))
+
         self.labels_ = -np.ones(len(X), dtype=np.int)
         offset = 0
 
-        for b in np.unique(blocks):
-            # Select data from block
-            mask = (blocks == b)
-            X_mask = X[mask, :]
-
-            if y is not None:
-                y_mask = y[mask]
-            else:
-                y_mask = None
-
-            if self.affinity == "precomputed":
-                X_mask = X_mask[:, mask]
-
-            # Fit a clusterer on the selected data
-            if len(X_mask) == 1:
-                clusterer = _SingleClustering()
-            elif self.fit_:
-                clusterer = clone(self.base_estimator)
-            elif self.partial_fit_:
-                if b in self.clusterers_:
-                    clusterer = self.clusterers_[b]
-                else:
-                    clusterer = clone(self.base_estimator)
-
-            if self.fit_ or not hasattr(clusterer, "partial_fit"):
-                try:
-                    clusterer.fit(X_mask, y=y_mask)
-                except TypeError:
-                    clusterer.fit(X_mask)
-            elif self.partial_fit_:
-                try:
-                    clusterer.partial_fit(X_mask, y=y_mask)
-                except TypeError:
-                    clusterer.partial_fit(X_mask)
-
-            self.clusterers_[b] = clusterer
-
-            # Save predictions
-            pred = np.array(clusterer.labels_)
-            mask_unknown = (pred == -1)
-            pred[~mask_unknown] += offset
-            self.labels_[mask] = pred
-            offset += np.max(clusterer.labels_) + 1
+        true_jobs = jobs_from_njobs(n_jobs)
+        unique_blocks = np.unique(blocks)
+        group_size = (len(unique_blocks) + true_jobs) / true_jobs
+
+        parallelizer = Parallel(n_jobs=n_jobs)
+        tasks_iterator = (delayed(_parallel_fit)(X, y, blocks, self.affinity,
+                                                 self.fit_,
+                                                 self.partial_fit_,
+                                                 self.base_estimator,
+                                                 self.clusterers_, b)
+                          for b in _grouper(group_size, unique_blocks))
+
+        result = parallelizer(tasks_iterator)
+
+        for bs in result:
+            for b, clusterer in bs:
+                # Save predictions
+                self.clusterers_[b] = clusterer
+                pred = np.array(clusterer.labels_)
+                mask_unknown = (pred == -1)
+                pred[~mask_unknown] += offset
+                self.labels_[(blocks == b)] = pred
+                offset += np.max(clusterer.labels_) + 1
 
         return self
 
-    def fit(self, X, y=None, blocks=None):
+    def fit(self, X, y=None, blocks=None, n_jobs=-1):
```

Comment:
Can you instead make `n_jobs` a parameter in the constructor? (and set it to `1` by default)

API-wise, the idea to have set all hyper-parameters at initialization. `fit` and other methods should only take input data as inputs -- i.e., nothing regarding how the execution is configured.